### ISSUE TITLE
CSS custom property set regression in selector-* rules - "cannot parse selector"

### ISSUE THREAD

**Issue by web-padawan** [stylelint/stylelint#2631]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A bug related to parsing of custom property sets like this appeared in 7.11.0.

> Which rule, if any, is this issue related to?

I'm not able to identify, sorry.

> What CSS is needed to reproduce this issue?

```css
body {

  --dot: {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  };
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "extends": [ "stylelint-config-standard" ]
}
```

> Which version of stylelint are you using?

`7.11.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

npm script. See [Glitch App](https://glitch.com/edit/#!/join/97898c1c-3eaa-4add-b470-8dad122af4cc) (press Logs button above the files list).

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, related to CSS mixins which are not yet standardized and [abandoned](http://www.xanthir.com/b4o00) by Tab Atkins.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The code above caused following output (not error):
```shell
 7:3     Cannot parse selector
```

PS: no such warnings with 7.10.1

**Comment by jeddy3** [comment]
@web-padawan Thanks for reporting and for using the template.

I can reproduce this regression. I believe it's caused by two things:

1. [We now throw](https://github.com/stylelint/stylelint/pull/1268/files) when `postcss-selector-parser` fails to parse a selector in the `selector-*` rules.
2. The CSS custom property set is being mistaken as a rule and `--dot:` is mistakenly parsed as a selector.

I believe the fix is to return early in all the affected `selector-*` rules if the rule [is not standard](https://github.com/stylelint/stylelint/blob/master/lib/utils/isStandardSyntaxRule.js). `no-descending-specificity` [does this right](https://github.com/stylelint/stylelint/blob/master/lib/rules/no-descending-specificity/index.js#L32-L34).

**Comment by evilebottnawi** [comment]
@jeddy3 i think be best just add `isCustom*` utils now, because `postcss` v6 handle them to another way.

**Comment by jeddy3** [comment]
Sounds good. Let's make use of `isStandardSyntaxRule`, rather than `isCustomPropertySet`, as that's consistent with how we do it in other `selector-*` rules like `selector-descendant-combinator-no-non-space`.

**Comment by evilebottnawi** [comment]
@web-padawan seems it is not valid use `custom properties sets` in `body`.

I try to add tests to `selector-attribute-brackets-space-inside` and works fine, seems it is buggy only one rule.

**Comment by web-padawan** [comment]
@evilebottnawi I just updated [Glitch app](https://glitch.com/edit/#!/join/97898c1c-3eaa-4add-b470-8dad122af4cc) above for more evidence. 
In our Polymer project we have quite a lot of mixins and it does not seem to depend on outer selector.

**Comment by jeddy3** [comment]
@evilebottnawi Does using `isStandardSyntaxRule` resolve the issue regardless of the outer selector? If so, then I think we're fine going with `isStandardSyntaxRule`. I think all the selector rules should be using `isStandardSyntaxRule` - I think it was just an oversight that they don't already.

**Comment by evilebottnawi** [comment]
@web-padawan 
Your right, from spec
```css
:root { --color: blue; }
div { --color: green; }
#alert { --color: red; }
* { color: var(--color); }

<p>I inherited blue from the root element!</p>
<div>I got green set directly on me!</div>
<div id='alert'>
  While I got red set directly on me!
  <p>I’m red too, because of inheritance!</p>
</div>
```
https://drafts.csswg.org/css-variables/#syntax

**Comment by evilebottnawi** [comment]
@jeddy3 i am working on this now, seems `isStandardSyntaxRule` is misleading, because `custom properties` and `custom property sets` are standard for css.

**Comment by jeddy3** [comment]
> @jeddy3 i am working on this now, seems isStandardSyntaxRule is misleading, because custom properties and custom property sets are standard for css.

The `isStandardSyntax*` utils are there to ignore constructs that PostCSS parses as a _particular_ CSS construct but aren't e.g.

```css
  --dot: {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  };
```

Is [parsed as a _rule_](https://astexplorer.net/#/gist/25eb06fae1f040ae94b434ca9204b1d8/1d3ded17b9608c0f5652e48d49bab4d7c257f981) by PostCSS, but it's not a rule - it's a custom property set. And `--dot:` isn't a _selector_, it's the _name of a custom property set_. 

The `isStandardSyntax*` utils allow us to filter out these, so we are sure we are only checking _actual_ rules and _actual_ selectors.

**Comment by web-padawan** [comment]
@evilebottnawi let me emphasize that, while `custom properties` ARE standard, `custom property sets` are NOT. Tab Atkins, who was a maintainer, of that [https://tabatkins.github.io/specs/css-apply-rule/](draft) now works on `::part()` spec as a replacement, which should specially target Shadow DOM,  as it is mentioned in the link I submitted above.

**Comment by evilebottnawi** [comment]
@jeddy3 do your think best place to solve `isStandardSyntaxRule`?

**Comment by evilebottnawi** [comment]
Also we don't ignore all rule
```css
body {
  --dot: {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  };

  span {
    color: red;
  }
}
```
For me best solution is `isStandardSyntaxSelector` for this place

**Comment by jeddy3** [comment]
> @jeddy3 do your think best place to solve is isStandardSyntaxRule?

Yes. I think I've tracked down the problem to `selector-pseudo-class-no-unknown`.

If [the node](https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-pseudo-class-no-unknown/index.js#L101-L103) is a `rule` one then we should check it is standard, with `isStandardSyntaxRule`, before continuing. Likewise, if the node is an `atRule` node then we should use `isStandardSyntaxAtRule`.

We use `isStandardSyntaxRule` is all the other `selector-*` rules, with the exception of the `selector-pseudo-class-whitelist` and `selector-pseudo-class-blacklist`. I think this is an oversight and these rules should also use `isStandardSyntaxRule` in addition to `isStandardSyntaxSelector`.

> Also we don't ignore all rule

`isStandardSyntaxRule` won't ignore all the rule as `root.walkRules()` iterates through rules regardless of nesting. Returning early will just ignore the custom property set and not the `span` rule.

**Comment by evilebottnawi** [comment]
@jeddy3 good solution, do it

**Comment by jeddy3** [comment]
Done in https://github.com/stylelint/stylelint/issues/2634
